src/work_queue.cpp

```diff
@@ -3,83 +3,157 @@
+  DataNode *node = consumer_list;
+  consumer_list = consumer_list->next;
+  --consumer_list_size;
+  lk.unlock();
+
+  data = node;
+  return true;
 }
 
-bool WorkQueue::peek(std::pair<int, queue_elm> &ret) {
-  do {
-    std::unique_lock<std::mutex> lk(rw_lock);
-    wq_empty.wait_for(lk, std::chrono::milliseconds(500), [this]{return (!empty() || no_block);});
-    if(!empty()) {
-      int temp = tail;
-      queue_array[tail].touched = true;
-      tail = incr(tail);
-      lk.unlock();
-
-      ret.first = temp;
-      ret.second = queue_array[temp];
-      return true;
-    }
+bool WorkQueue::peek_batch(std::vector<DataNode *> &node_vec, int batch_size) {
+  assert(batch_size <= len); // cannot request a batch bigger than the work queue
+
+  node_vec.clear(); // clear out any old data
+  node_vec.reserve(batch_size);
+
+  // wait until consumer queue is large enough
+  std::unique_lock<std::mutex> lk(consumer_list_lock);
+  consumer_condition.wait(lk, 
+    [this, batch_size]{return consumer_list_size >= batch_size || non_block;});
+
+  // printf("WQ: Peek-batch\n");
+  // print();
+
+  if (empty()) {
     lk.unlock();
-  }while(!no_block);
-  return false;
+    return false;
+  }
+
+  // pull data from head of consumer_list
+  for(int i = 0; i < batch_size; i++) {
+    if (consumer_list == nullptr) break; // if non_block is true may not be able to get full batch
+
+    node_vec.push_back(consumer_list);
+    consumer_list = consumer_list->next;
+    --consumer_list_size;
+  }
+  lk.unlock();
+  return true;
+}
+
+void WorkQueue::peek_callback(DataNode *node) {
+  producer_list_lock.lock();
+  // printf("WQ: Callback\n");
+  // print();
+  node->next = producer_list;
+  producer_list = node;
+  producer_list_lock.unlock();
+  producer_condition.notify_one();
+  // printf("WQ: Callback done\n");
 }
 
-void WorkQueue::pop(int i) {
-  rw_lock.lock();
-  queue_array[i].dirty   = false; // this data has been processed and this slot may now be overwritten
-  queue_array[i].touched = false; // may read this slot
-  rw_lock.unlock();
-  wq_full.notify_one();
+void WorkQueue::set_non_block(bool _block) {
+  consumer_list_lock.lock();
+  non_block = _block;
+  consumer_list_lock.unlock();
+  consumer_condition.notify_all();
 }
 
 void WorkQueue::print() {
-  printf("WQ: head=%i, tail=%i, is_full=%s, is_empty=%s\n", 
-    head, tail, full()? "true" : "false", empty()? "true" : "false");
+  std::string to_print = "";
+
+  int p_size = 0;
+  DataNode *temp = producer_list;
+  while (temp != nullptr) {
+    to_print += std::to_string(p_size) + ": " + std::to_string((uint64_t)temp) + "\n";
+    temp = temp->next;
+    ++p_size;
+  }
+  int c_size = 0;
+  temp = consumer_list;
+  while (temp != nullptr) {
+    to_print += std::to_string(c_size) + ": " + std::to_string((uint64_t)temp) + "\n";
+    temp = temp->next;
+    ++c_size;
+  }
+  printf("WQ: producer_queue size = %i consumer_queue size = %i\n%s", p_size, c_size, to_print.c_str());
```

Comment:
Consider holding the lock here

Reply:
This function is generally called from within a lock. It's also just for debugging.